### PR TITLE
Added diagnostic check for watertight nodesets

### LINKED ISSUE #28822: Add diagnostic check for water-tight nodesets
### Motivation

Adding more features to the mesh diagnostic system will help users debug issues with their meshes.  

Linked to #25278 and #28816 

### Design

Should be added to MeshDiagnostics Generator. There is a diagnostic feature in progress that would detect when a mesh has external sides without a sideset. This would be similar, but instead look for missing nodesets instead of sidesets.

### Impact

Makes it easier to find issues with user meshes.

### PR DESCRIPTION
**Pull request by DanielYankura**
-Loops through all elements and sides
-If side is external it checks if all its nodes are assigned to a nodeset

closes #28822

## Reason

Adds new diagnostic tool for users

## Design

If feature it turned on, it loops through each element, then the element's sides. If the sides are external it checks each node to ensure it's been assigned to a nodeset. If not, the node id is printed for reference.

## Impact

Adds watertight nodeset check to Mesh Diagnostics

### CODE REVIEW

File: framework/src/meshgenerators/MeshDiagnosticsGenerator.C
Comment (GiudGiud) [inline framework/src/meshgenerators/MeshDiagnosticsGenerator.C:52]: let s add a parameter for the nodesets that are supposed to englobe the entire mesh. when this parameter is in use, the nodes are only checked against those "blessed" nodesets

same for sidsets
actually now that I see that this is not an option for sidesets, let s keep it for future work

Comment (DanielYankura) [reply framework/src/meshgenerators/MeshDiagnosticsGenerator.C:52]: Sounds good. I'll open a new issue for adding that parameter to the sideset and nodeset checks.

File: framework/src/meshgenerators/MeshDiagnosticsGenerator.C
Comment (GiudGiud) [inline framework/src/meshgenerators/MeshDiagnosticsGenerator.C:362]: where is the _boundary_node_id being initialized in libmesh? is the previous call doing it?

File: framework/src/meshgenerators/MeshDiagnosticsGenerator.C
Comment (GiudGiud) [inline framework/src/meshgenerators/MeshDiagnosticsGenerator.C:368]:
```suggestion
    for (const auto i : elem->side_index_range())
```

File: framework/src/meshgenerators/MeshDiagnosticsGenerator.C
Comment (GiudGiud) [inline framework/src/meshgenerators/MeshDiagnosticsGenerator.C:373]:
```suggestion
        // Side is external, now check nodes
```

File: framework/src/meshgenerators/MeshDiagnosticsGenerator.C
Comment (GiudGiud) [inline framework/src/meshgenerators/MeshDiagnosticsGenerator.C:391]:
```suggestion
                        " is on an external boundary of the mesh, but has not been assigned to a nodeset";
```

File: test/tests/meshgenerators/mesh_diagnostics_generator/tests
Comment (GiudGiud) [inline test/tests/meshgenerators/mesh_diagnostics_generator/tests:54]:
```suggestion
      detail = '2D meshed square with one external node that does not belong to a nodeset'
```

File: test/tests/meshgenerators/mesh_diagnostics_generator/tests
Comment (GiudGiud) [inline test/tests/meshgenerators/mesh_diagnostics_generator/tests:62]: same

**Review by GiudGiud — COMMENTED** [review]
looks good

File: framework/src/meshgenerators/MeshDiagnosticsGenerator.C
Comment (DanielYankura) [reply framework/src/meshgenerators/MeshDiagnosticsGenerator.C:362]: So I looked into it. The node ids get added to _boundary_node_id after execute_mesh_generators is called in ActionWarehouse.C. From what I can tell the mesh is always initialized before this test could ever be called.

Comment (GiudGiud) [reply framework/src/meshgenerators/MeshDiagnosticsGenerator.C:362]: >  after execute_mesh_generators

wouldnt that be too late then?

Comment (DanielYankura) [reply framework/src/meshgenerators/MeshDiagnosticsGenerator.C:362]: Sorry, it's the execute_mesh_generators task that initializes _boundary_node_id. After initialization mesh diagnostics runs. So that task runs both I believe, but it should initialize first.

File: framework/src/meshgenerators/MeshDiagnosticsGenerator.C
Comment (DanielYankura) [reply framework/src/meshgenerators/MeshDiagnosticsGenerator.C:386]: Thanks! I was trying to figure out why it was failing that test. I think I have the same issue in #28701 so I'll fix it in that one as well.

Comment (GiudGiud) [reply framework/src/meshgenerators/MeshDiagnosticsGenerator.C:386]: I didnt remember that one was still open. Do you need anything else from me for that? Are we waiting on Roy's review?

Comment (DanielYankura) [reply framework/src/meshgenerators/MeshDiagnosticsGenerator.C]: I think I'm just waiting on Roy's review. Though if this one get's merged first I might have to make some changes to the other before it can be merged.

**Review by GiudGiud — APPROVED** [review]

File: framework/src/meshgenerators/MeshDiagnosticsGenerator.C
Comment (roystgnr) [inline framework/src/meshgenerators/MeshDiagnosticsGenerator.C:361]: What's going on here?  This is the `build_side_list()` variant that returns a vector of tuples, but if nothing uses the return value then it's just an expensive no-op.

File: framework/src/meshgenerators/MeshDiagnosticsGenerator.C
Comment (roystgnr) [inline framework/src/meshgenerators/MeshDiagnosticsGenerator.C:382]: Could we just test `boundary_info.n_boundary_ids(node)` here?  The raw `get_nodeset_map()` accessor was an encapsulation mistake I should never have approved, and I still vaguely dream of deprecating it someday.

**Review by roystgnr — CHANGES_REQUESTED** [review]
In principle, I oppose the very concept of checking for watertight nodesets, and would be happy to rant at length about why.  (nobody else would be happy, so I won't)

In practice, I recognize that some MOOSE workflows will be using nodesets as pseudo-sidesets until the end of time; this is a great way for users of those workflows to check for meshing problems.  I just have one minor problem and one minor request here.

File: framework/src/meshgenerators/MeshDiagnosticsGenerator.C
Comment (DanielYankura) [reply framework/src/meshgenerators/MeshDiagnosticsGenerator.C:361]: I think I copied that over from a similar function I made previously and then forgot to remove it. I'l fix that real quick.

File: framework/src/meshgenerators/MeshDiagnosticsGenerator.C
Comment (DanielYankura) [reply framework/src/meshgenerators/MeshDiagnosticsGenerator.C:382]: Just want to check, would that return true if the node is assigned to a sideset but not a nodeset?

**Review by roystgnr — APPROVED** [review]

**Pull request merged**
